RecoEgamma/EgammaTools/src/EnergyScaleCorrection.cc

```diff
@@ -0,0 +1,426 @@
+      else phi = std::stod(phiString);
+      
+      if(errPhiString=="M_PI_2") errPhi=M_PI_2;
+      else errPhi = std::stod(errPhiString);
+      
+      addSmearing(category, runMin, runMax, rho,  errRho, phi, errPhi, eMean, errEMean);
+      
+    } else {
+      file >> category >> rho >> phi;
+      errRho = errPhi = eMean = errEMean = 0;
+      addSmearing(category, runMin, runMax, rho,  errRho, phi, errPhi, eMean, errEMean);
+    }
+
+  }
+  
+  file.close();
+  return;
+}
+
+std::ostream& EnergyScaleCorrection::ScaleCorrection::print(std::ostream& os)const
+{
+  os <<  "( "<< scale_ << " +/- " << scaleErrStat_ << " +/- " << scaleErrSyst_ << " +/- " << scaleErrGain_ <<")" ;
+  return os; 
+}
+
+float EnergyScaleCorrection::ScaleCorrection::scaleErr(const std::bitset<kErrNrBits>& uncBitMask)const
+{
+  double totErr(0);
+  auto pow2 = [](const double& x){return x*x;};
+  
+  if(uncBitMask.test(kErrStatBitNr)) totErr+=pow2(scaleErrStat_);
+  if(uncBitMask.test(kErrSystBitNr)) totErr+=pow2(scaleErrSyst_);  
+  if(uncBitMask.test(kErrGainBitNr)) totErr+=pow2(scaleErrGain_);
+  
+  return std::sqrt(totErr);
+}
+
+std::ostream& EnergyScaleCorrection::SmearCorrection::print(std::ostream& os)const
+{
+  os << rho_ << " +/- " << rhoErr_ 
+     <<  "\t"
+     << phi_ << " +/- " << phiErr_
+     <<  "\t"
+     << eMean_ << " +/- " << eMeanErr_;
+  return os; 
+}
+
+//here be dragons
+//this function is nasty and needs to be replaced
+EnergyScaleCorrection::CorrectionCategory::CorrectionCategory(const std::string& category,int runnrMin,int runnrMax):
+  runMin_(runnrMin),runMax_(runnrMax),etaMin_(0),etaMax_(3),
+  r9Min_(-1),r9Max_(999),etMin_(0),etMax_(9999999),gain_(0)
+{
+  size_t p1, p2; // boundary
+
+  // eta region
+  p1 = category.find("absEta_");
+  if(category.find("absEta_0_1") != std::string::npos) {
+    etaMin_ = 0;
+    etaMax_ = 1;
+  } else if(category.find("absEta_1_1.4442") != std::string::npos) {
+    etaMin_ = 1;
+    etaMax_ = 1.479;
+  }
+  else if(category.find("absEta_1.566_2") != std::string::npos) {
+    etaMin_ = 1.479;
+    etaMax_ = 2;
+  }
+  else if(category.find("absEta_2_2.5") != std::string::npos) {
+    etaMin_ = 2;
+    etaMax_ = 3;
+  } else {
+    if(p1 != std::string::npos) {
+      p1 = category.find("_", p1);
+      p2 = category.find("_", p1 + 1);
+      etaMin_ = std::stof(category.substr(p1 + 1, p2 - p1 - 1));
+      p1 = p2;
+      p2 = category.find("-", p1);
+      etaMax_ = std::stof(category.substr(p1 + 1, p2 - p1 - 1));
+    }
+  }
+  
+  if(category.find("EBlowEta") != std::string::npos) {
+    etaMin_ = 0;
+    etaMax_ = 1;
+  };
+  if(category.find("EBhighEta") != std::string::npos) {
+    etaMin_ = 1;
+    etaMax_ = 1.479;
+  };
+  if(category.find("EElowEta") != std::string::npos) {
+    etaMin_ = 1.479;
+    etaMax_ = 2;
+  };
+  if(category.find("EEhighEta") != std::string::npos) {
+    etaMin_ = 2;
+    etaMax_ = 7;
+  };
+  
+  // Et region
+  p1 = category.find("-Et_");
+  p2 = p1 + 1;
```

Comment:
static analyzer flagged this line
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22531/26791/llvm-analysis/report-a08295.html#EndPath
please check

Reply:
line 340 can be removed

Reply:
removed